mitmproxy/proxy/mode.py

```diff
@@ -0,0 +1,173 @@
+        """Validation of data happens here."""
+
+    @property
+    @abstractmethod
+    def transport_protocol(self) -> Literal["tcp", "udp"]:
+        """The transport protocol used by this mode's server."""
+
+    @classmethod
+    @cache
+    def parse(cls: Type[Self], spec: str) -> Self:
+        """
+        Parse a proxy mode specification and return the corresponding `ProxyMode` instance.
+        """
+        head, _, listen_at = spec.rpartition("@")
+        if not head:
+            head = listen_at
+            listen_at = ""
+
+        mode, _, data = head.partition(":")
+
+        if listen_at:
+            if ":" in listen_at:
+                host, _, port_str = listen_at.rpartition(":")
+            else:
+                host = None
+                port_str = listen_at
+            try:
+                port = int(port_str)
+                if port < 0 or 65535 < port:
+                    raise ValueError
+            except ValueError:
+                raise ValueError(f"invalid port: {port_str}")
+        else:
+            host = None
+            port = None
+
+        try:
+            mode_cls = ProxyMode.__types[mode.lower()]
+        except KeyError:
+            raise ValueError(f"unknown mode")
+
+        if not issubclass(mode_cls, cls):
+            raise ValueError(f"{mode!r} is not a spec for a {cls.type} mode")
+
+        return mode_cls(
+            full_spec=spec,
+            data=data,
+            custom_listen_host=host,
+            custom_listen_port=port
+        )
+
+    def listen_host(self, default: str | None = None) -> str:
+        """
+        Return the address a server for this mode should listen on. This can be either directly
+        specified in the spec or taken from a user-configured global default (`options.listen_host`).
+        By default, return an empty string to listen on all hosts.
+        """
+        if self.custom_listen_host is not None:
+            return self.custom_listen_host
+        elif default is not None:
+            return default
+        else:
+            return ""
+
+    def listen_port(self, default: int | None = None) -> int:
+        """
+        Return the port a server for this mode should listen on. This can be either directly
+        specified in the spec, taken from a user-configured global default (`options.listen_port`),
+        or from `ProxyMode.default_port`.
+        """
+        if self.custom_listen_port is not None:
+            return self.custom_listen_port
+        elif default is not None:
+            return default
+        else:
+            return self.default_port
+
+    @classmethod
+    def from_state(cls, state):
+        return ProxyMode.parse(state)
+
+    def get_state(self):
+        return self.full_spec
+
+    def set_state(self, state):
+        if state != self.full_spec:
+            raise RuntimeError("Proxy modes are frozen.")
+
+
+class AsyncioProxyMode(ProxyMode, metaclass=ABCMeta):
```

Comment:
I feel this complicates things more than it helps. Maybe we can move `description` into ProxyMode itself and then keep the layer information as part of individual server instance subclasses? That avoids the special-casing in `ServerInstance.make`, keeps all modes as immediate subclasses of ProxyMode,  retains a dumb 1-1 mapping between modes and instances, and avoids any circular spec -> layer -> spec dependencies. This may entirely be "code I wrote myself is easier to understand" bias, so feel free to go ahead with this approach if you feel it's more straightforward. Just a subjective impression. :)

Reply:
I guess it's just my code-compressing attitude, but you're right. Having 1:1 `ServerInstance`s removes all the `if TYPE_CHECKING` stuff, so I'll undo the changes.